pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -210,35 +224,57 @@ func (c *Controller) Apply(change *mcpclient.Change) error {
 		}
 	}
 
-	// add/update
-	for namespace, byName := range innerStore {
-		for name, config := range byName {
-			if prevByNamespace, ok := prevStore[namespace]; ok {
-				if prevConfig, ok := prevByNamespace[name]; ok {
-					if config.ResourceVersion != prevConfig.ResourceVersion {
-						dispatch(config, model.EventUpdate)
-					}
-				} else {
-					dispatch(config, model.EventAdd)
-				}
+	// update the internal store and optionally generate add/update/delete events
+	for _, obj := range addedAndUpdate {
+		byType, ok := c.configStore[obj.Type]
+		if !ok {
+			byType = make(map[string]map[string]*model.Config)
+			c.configStore[obj.Type] = byType
+		}
+
+		byNamespace, ok := byType[obj.Namespace]
+		if !ok {
+			byNamespace = make(map[string]*model.Config)
+			byType[obj.Namespace] = byNamespace
+		}
+
+		if generateAllEvents {
+			if _, ok := c.configStore[obj.Type][obj.Namespace][obj.Name]; ok {
+				dispatch(*obj, model.EventUpdate)
 			} else {
-				dispatch(config, model.EventAdd)
+				dispatch(*obj, model.EventAdd)
 			}
 		}
+		c.configStore[obj.Type][obj.Namespace][obj.Name] = obj
 	}
+	for _, remove := range change.Removed {
+		namespace, name := extractNameNamespace(remove)
 
-	// remove
-	for namespace, prevByName := range prevStore {
-		for name, prevConfig := range prevByName {
-			if byNamespace, ok := innerStore[namespace]; ok {
-				if _, ok := byNamespace[name]; !ok {
-					dispatch(prevConfig, model.EventDelete)
-				}
-			} else {
-				dispatch(prevConfig, model.EventDelete)
+		typ := descriptor.Type
+		if namespace == "" && descriptor.Type == model.AuthenticationPolicy.Type {
+			typ = model.AuthenticationMeshPolicy.Type
+		}
+
+		if obj, ok := c.configStore[typ][namespace][name]; ok {
+			if generateAllEvents {
+				dispatch(*obj, model.EventDelete)
 			}
+		} else {
+			log.Warnf("Removing unknown resource type=%s name=%s", typ, remove)
+		}
+
+		delete(c.configStore[typ][namespace], name)
+		if len(c.configStore[typ][namespace]) == 0 {
+			delete(c.configStore[typ], namespace)
 		}
 	}
+	c.configStoreMu.Unlock()
+
+	// TODO replace with indirect call to DiscoveryServer::ConfigUpdate()
+	if !generateAllEvents {
```

Comment:
@ayj Should I go ahead and close this [PR](https://github.com/istio/istio/pull/10446#pullrequestreview-190088287) since it's covered here?

Reply:
I think so. This PR still has some TODO's around optimizing ServiceEntry updates and invoking ConfigUpdate() directly that we'll still need to address in a follow-up PR